### PR TITLE
[xray] raylet scheduling mechanism with a simple spillback policy

### PR DESCRIPTION
**Pull request by atumanov**
## What do these changes do?
* distribute load and resource information on a heartbeat
* for each raylet, maintain total and available resource capacity as well as measure of current load
* this PR introduces a new notion of load, defined as a sum of all resource demand induced by queued ready tasks on the local raylet. This provides a heterogeneity-aware measure of load that supersedes legacy Ray's task count as a proxy for load.
* modify the scheduling policy to perform *capacity-based*, *load-aware*, *optimistically concurrent* resource allocation
* perform task spillover to the heartbeating node in response to a heartbeat, implementing  heterogeneity-aware late-binding/work-stealing.

### CODE REVIEW

File: src/ray/raylet/scheduling_resources.h [hunk 1]
```diff
@@ -80,18 +80,26 @@ class ResourceSet {
...
   /// \return True if the resource set was added successfully. False otherwise.
-  bool AddResources(const ResourceSet &other);
+  bool AddResourcesStrict(const ResourceSet &other);
 
-  /// \brief Subtract a set of resources from the current set of resources.
+  /// \brief Aggregate resources from the other set into this set, adding any missing
+  /// resource labels to this set.
+  ///
+  /// \param other: The other resource set to add.
+  void AddResources(const ResourceSet &other);
```
Comment (robertnishihara) [inline src/ray/raylet/scheduling_resources.h]: `\return Void.`

File: src/ray/raylet/scheduling_policy.h [hunk 2]
```diff
@@ -22,15 +22,20 @@ class SchedulingPolicy {
...
+  /// producing a mapping of tasks to raylets.
   ///
-  ///  \param cluster_resources: a set of cluster resources representing
-  ///         configured and current resource capacity on each node.
-  /// \return Scheduling decision, mapping tasks to node managers for placement.
+  /// \param cluster_resources: a set of cluster resources containing resource and load
+  /// information for some subset of the cluster. For all client IDs in the returned
+  /// placement map, the corresponding SchedulingResources::resources_load_ is
+  /// incremented by the aggregate resource demand of the tasks assigned to it.
+  ///
```
Comment (robertnishihara) [inline src/ray/raylet/scheduling_policy.h]: there shouldn't be an extra `///` here

File: src/ray/raylet/scheduling_policy.h [hunk 3]
```diff
@@ -22,15 +22,20 @@ class SchedulingPolicy {
...
+  /// \return Scheduling decision, mapping tasks to raylets for placement.
   std::unordered_map<TaskID, ClientID> Schedule(
-      const std::unordered_map<ClientID, SchedulingResources> &cluster_resources,
-      const ClientID &local_client_id, const std::vector<ClientID> &others);
+      std::unordered_map<ClientID, SchedulingResources> &cluster_resources,
+      const ClientID &local_client_id);
```
Comment (robertnishihara) [inline src/ray/raylet/scheduling_policy.h]: document `local_client_id`

File: src/ray/raylet/scheduling_policy.h [hunk 4]
```diff
@@ -22,15 +22,20 @@ class SchedulingPolicy {
   /// \return Void.
   SchedulingPolicy(const SchedulingQueue &scheduling_queue);
 
-  /// Perform a scheduling operation, given a set of cluster resources and
-  /// producing a mapping of tasks to node managers.
+  /// \brief Perform a scheduling operation, given a set of cluster resources and
+  /// producing a mapping of tasks to raylets.
```
Comment (robertnishihara) [inline src/ray/raylet/scheduling_policy.h]: This no longer produces a "mapping"

File: src/ray/raylet/node_manager.cc [hunk 5]
```diff
@@ -506,6 +533,14 @@ void NodeManager::ProcessClientMessage(
...
     // Return the worker to the idle pool.
     worker_pool_.PushWorker(std::move(worker));
+    // Local resource availability changed: invoke scheduling policy for local node.
+    const ClientID &local_client_id = gcs_client_->client_table().GetLocalClientId();
+    cluster_resource_map_[local_client_id].SetLoadResources(
+        local_queues_.GetResourceLoad());
+    std::unordered_map<ClientID, SchedulingResources> local_resource_map(
+        {{local_client_id, cluster_resource_map_[local_client_id]}});
+    // Invoke the scheduling policy only on local resources.
+    // ScheduleTasks(local_resource_map);
```
Comment (robertnishihara) [inline src/ray/raylet/node_manager.cc]: Remove this and the comment

**Comment by robertnishihara** [comment]
We should also update the diagram in https://github.com/ray-project/ray/blob/master/src/ray/raylet/design_docs/task_states.rst.

**Comment by atumanov** [comment]
Jenkins retest this please

**Review by robertnishihara — APPROVED** [review]

**Pull request merged**